rust/src/pgsql/pgsql.rs

```diff
@@ -0,0 +1,820 @@
+            if !probe_tc(input) {
+                // Out of sync, we'll just try again next time.
+                SCLogDebug!("Suricata interprets there's a gap in the response");
+                return AppLayerResult::ok();
+            }
+
+            // It seems we're in sync with a message header, clear gap state and keep parsing.
+            self.response_gap = false;
+        }
+
+        let mut start = input;
+        while start.len() > 0 {
+            match PgsqlState::state_based_resp_parsing(self.state_progress, start) {
+                Ok((rem, response)) => {
+                    start = rem;
+                    SCLogDebug!("Response is {:?}", &response);
+                    if let Some(state) = self.response_get_next_state(&response, flow) {
+                        self.state_progress = state;
+                    };
+                    if let Some(tx) = self.find_or_create_tx() {
+                        tx.responses.push(response);
+                    } else {
+                        // If there isn't a new transaction, we'll consider Suri should move on
+                        return AppLayerResult::ok();
+                    };
+                }
+                Err(nom::Err::Incomplete(_needed)) => {
+                    let consumed = input.len() - start.len();
+                    let needed_estimation = start.len() + 1;
+                    SCLogDebug!(
+                        "Needed: {:?}, estimated needed: {:?}, start is {:?}",
+                        _needed,
+                        needed_estimation,
+                        &start
+                    );
+                    return AppLayerResult::incomplete(consumed as u32, needed_estimation as u32);
+                }
+                Err(_) => {
+                    SCLogDebug!("Error while parsing PostgreSQL response");
+                    return AppLayerResult::err();
+                }
+            }
+        }
+
+        // All input was fully consumed.
+        return AppLayerResult::ok();
+    }
+
+    fn tx_iterator(
+        &mut self, min_tx_id: u64, state: &mut u64,
+    ) -> Option<(&PgsqlTransaction, u64, bool)> {
+        let mut index = *state as usize;
+        let len = self.transactions.len();
+
+        while index < len {
+            let tx = &self.transactions[index];
+            if tx.tx_id < min_tx_id + 1 {
+                index += 1;
+                continue;
+            }
+            *state = index as u64;
+            return Some((tx, tx.tx_id - 1, (len - index) > 1));
+        }
+
+        return None;
+    }
+
+    fn on_request_gap(&mut self, _size: u32) {
+        self.request_gap = true;
+    }
+
+    fn on_response_gap(&mut self, _size: u32) {
+        self.response_gap = true;
+    }
+}
+
+/// Probe for a valid PostgreSQL request
+///
+/// PGSQL messages don't have a header per se, so we parse the slice for an ok()
+fn probe_ts(input: &[u8]) -> bool {
+    SCLogDebug!("We are in probe_ts");
```

Comment:
if we have a gap in the stream, that can result in a temporary incomplete result, is that so? If it is, then I think it could happen. But I must probably investigate this further.

Reply:
> if we have a gap in the stream, that can result in a temporary incomplete result, is that so?

Not exactly.
Rather if the client sends the TCP data one byte at a time in each packet...

Reply:
Hm... I haven't found mentions of that, so far, in the documentation. Could you recommend a good way for one to verify this?

Reply:
cf `rs_http2_probing_parser_tc` returning either `ALPROTO_FAILED`, `ALPROTO_UNKNOWN` (for incomplete) or `ALPROTO_HTTP2`

Reply:
Ok, thanks :)